contracts/expiring_plugin.algo.ts

```diff
@@ -0,0 +1,121 @@
+    newExpiringPlugin(delegatedApp: Application, duration: uint64, authAddr: Address, allowedApp: Application, method: bytes): void {
+        // mint the expiring plugin contract
+        sendMethodCall<[uint64, Address, Application, bytes], void>({
+            name: 'createApplication',
+            approvalProgram: ExpiringPlugin.approvalProgram(),
+            clearStateProgram: ExpiringPlugin.clearProgram(),
+            globalNumUint: 3,
+			globalNumByteSlice: 2,
+            methodArgs: [duration, authAddr, allowedApp, method],
+            fee: 0,
+        });
+
+        const child = this.itxn.createdApplicationID;
+        
+        sendMethodCall<[Application], void>({
+            sender: delegatedApp.address,
+            name: 'addPlugin',
+            applicationID: delegatedApp,
+            methodArgs: [child],
+        })
+    }
+}
+
+
+/**
+ * ExpiringPlugin is a dynamic general purpose expiring plugin.
+ * It allows you to delegate a specific address the ability 
+ * to call a single method on a specific other contract for a limited time.
+ */
+export class ExpiringPlugin extends Contract {
+    programVersion = 10;
+
+    /**
+     * The time at which this contract was created
+     */
+    createdAt = GlobalStateKey<uint64>();
+
+    /**
+     * The duration is valid for
+     */
+    duration = GlobalStateKey<uint64>();
+
+    /**
+     * The address that is authorized to call the method
+     */
+    authAddr = GlobalStateKey<Address>();
+
+    /**
+     * The application that is authorized to be called
+     */
+    allowedApp = GlobalStateKey<Application>();
+
+    /**
+     * The method that is authorized to be called
+     */
+    method = GlobalStateKey<bytes>();
+
+    /**
+     * set the all state values at mint
+     * @param duration 
+     * @param authAddr 
+     * @param allowedApp 
+     * @param method 
+     */
+    createApplication(duration: uint64, authAddr: Address, allowedApp: Application, method: bytes): void {
+        this.createdAt.value = globals.latestTimestamp;
+        this.duration.value = duration;
+        this.authAddr.value = authAddr;
+        this.allowedApp.value = allowedApp;
+        this.method.value = method;
+    }
+
+    /**
+     * verifyAppCall ensures the external app call being made
+     * is valid and the contract hasn't expired
+     */
+    verify(sender: Account, appCall: AppCallTxn): void {
+        // ensure this app call is still valid in regards to the timeout
+        const expiresAt = this.createdAt.value + this.duration.value;
+        assert(
+            !(globals.latestTimestamp > expiresAt),
+            // ensure the sender is the authAddr
+            this.txn.sender === this.authAddr.value,
+            // ensure the method being called is the correct method
+            appCall.applicationArgs[0] === method(this.method.value)
+        )
+
+        // ensure we're calling the correct app from the correct account
+        verifyAppCallTxn(appCall, {
```

Comment:
Rather than verifying the app call I think we would need to rekey to the app here. Otherwise we need to trust the the app call is only callable when `verify` is called